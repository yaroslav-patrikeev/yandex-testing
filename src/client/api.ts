import axios from 'axios';
import { CartState, CheckoutFormData, CheckoutResponse, Product, ProductShortInfo } from '../common/types';

const currentBugId = Number(process.env.BUG_ID) || undefined;

export class ExampleApi {
    constructor(private readonly basename: string) {

    }

    async getProducts() {
        return await axios.get<ProductShortInfo[]>(`${this.basename}/api/products`, { params: { bug_id: currentBugId } });
    }

    async getProductById(id: number) {
        return await axios.get<Product>(`${this.basename}/api/products/${id}`, { params: { bug_id: currentBugId } });
    }

    async checkout(form: CheckoutFormData, cart: CartState) {
        return await axios.post<CheckoutResponse>(`${this.basename}/api/checkout`, { form, cart }, { params: { bug_id: currentBugId } });
    }
}

export const LOCAL_STORAGE_CART_KEY = 'example-store-cart';

export class CartApi {
    getState(): CartState {
        try {
            const json = localStorage.getItem(LOCAL_STORAGE_CART_KEY);
            return JSON.parse(json) as CartState || {};
        } catch {
            return {};
        }
    }

    setState(cart: CartState) {
        localStorage.setItem(LOCAL_STORAGE_CART_KEY, JSON.stringify(cart));
    }
}
