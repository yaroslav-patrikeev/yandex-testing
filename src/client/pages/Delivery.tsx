import React from "react";
import { Helmet } from "react-helmet";
import { cn } from "@bem-react/classname";

import { Image } from "../components/Image";

const bem = cn("Delivery");

export const Delivery: React.FC = () => {
  return (
    <div className={bem()}>
      <Helmet title="Delivery" />
      <div className="row">
        <div className="col">
          <h1>Delivery</h1>
          <p>
            Swift and Secure Delivery: Experience the convenience of hassle-free
            shipping with our scratchers. We understand the excitement of
            receiving your new cat furniture, so we prioritize swift delivery to
            your doorstep. Rest assured, your order is handled with care every
            step of the way, ensuring it arrives safely and securely.
          </p>
          <Image className="w-25 mb-4" />
          <p>
            Track Your Package with Ease: Stay informed and in control of your
            delivery with our easy-to-use tracking system. From the moment your
            order is placed to the minute it reaches your home, you can monitor
            its journey in real-time. No more guessing games â€“ know exactly when
            to expect your package and plan accordingly.
          </p>
          <p>
            Customer Satisfaction Guaranteed: Your satisfaction is our top
            priority, which is why we go above and beyond to provide exceptional
            delivery service. If you have any questions or concerns about your
            shipment, our dedicated customer support team is here to assist you
            every step of the way. Trust us to deliver not only your scratcher
            but also peace of mind.
          </p>
        </div>
      </div>
    </div>
  );
};
