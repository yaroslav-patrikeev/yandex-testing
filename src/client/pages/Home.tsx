import React from "react";
import { Helmet } from "react-helmet";
import { cn } from "@bem-react/classname";

const bem = cn("Home");

export const Home: React.FC = () => {
  return (
    <div className={bem()}>
      <Helmet title="Welcome" />
      <div className="row">
        <div className="col bg-orange text-white py-4 bg-opacity-75">
          <p className="display-3">Welcome to Kogtetochka store!</p>
          <p className="lead">
            We have a large assortment of scratching posts!
          </p>
        </div>
      </div>
      <div className="row mb-4">
        <div className="col-12 col-md-4 bg-light py-3">
          <h1>Stability</h1>
          <p className="lead">
            Our scratching posts are crafted with precision and designed for
            unparalleled stability. Made from high-quality materials, they
            provide a sturdy platform for your cat's scratching needs.
          </p>
        </div>
        <div className="col-12 col-md-4 bg-light py-3">
          <h1>Comfort</h1>
          <p className="lead">
            Pamper your feline friend with the luxurious comfort of our
            scratching posts. Covered in soft, plush fabric, they offer a cozy
            retreat for your cat to relax and unwind.
          </p>
        </div>
        <div className="col-12 col-md-4 bg-light py-3">
          <h1>Design</h1>
          <p className="lead">
            Engage your cat's natural instincts and keep them entertained for
            hours with our interactive scratching posts. Featuring built-in toys
            and enticing textures, they stimulate your cat's senses and
            encourage active play.
          </p>
        </div>
      </div>
      <div className="row mb-4">
        <div className="col-12py-3">
          <p className="fs-1">
            Empower Your Coding Journey with Every Scratch â€“ Get Your Paws on
            Our Purr-fect Scratchers Today!
          </p>
        </div>
      </div>
    </div>
  );
};
