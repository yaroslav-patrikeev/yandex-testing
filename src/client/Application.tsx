import React, { FC, useCallback, useState } from 'react';
import { Link, NavLink, Route, Routes } from 'react-router-dom';
import { cn } from '@bem-react/classname';

import { Home } from './pages/Home';
import { Catalog } from './pages/Catalog';
import { Product } from './pages/Product';
import { Delivery } from './pages/Delivery';
import { Contacts } from './pages/Contacts';
import { Helmet } from 'react-helmet';
import { Cart } from './pages/Cart';
import { useSelector } from 'react-redux';
import { ApplicationState } from './store';

const bem = cn('Application');

export const Application: FC = () => {
    const [collapsed, setCollapsed] = useState(true);
    const cart = useSelector((s: ApplicationState) => s.cart);

    const toggle = useCallback(() => setCollapsed(!collapsed), [setCollapsed, collapsed]);
    const hide = useCallback(() => {
        if (process.env.BUG_ID === '4') {
            setCollapsed(false);
        } else {
            setCollapsed(true);
        }
    }, [setCollapsed]);

    const count = Object.keys(cart).length;
    const cartLabel = count ? `Cart (${count})` : 'Cart';
    const navbarClass = collapsed ? 'collapse navbar-collapse' : 'navbar-collapse';

    return (
        <div className={bem()}>
            <Helmet titleTemplate='%s â€” Kogtetochka store' />
            <nav className='navbar navbar-expand-sm navbar-light bg-light'>
                <div className='container'>
                    <Link className={bem('Brand', ['navbar-brand'])} to='/'>
                        Kogtetochka store
                    </Link>
                    <button className={bem('Toggler', ['navbar-toggler'])} aria-label='Toggle navigation' onClick={toggle}>
                        <span className='navbar-toggler-icon'></span>
                    </button>
                    <div className={bem('Menu', [navbarClass])}>
                        <div className='navbar-nav'>
                            <NavLink className={({ isActive }) => (isActive ? 'active nav-link' : 'nav-link')} to='/catalog' onClick={hide}>
                                Catalog
                            </NavLink>
                            <NavLink className={({ isActive }) => (isActive ? 'active nav-link' : 'nav-link')} to='/delivery' onClick={hide}>
                                Delivery
                            </NavLink>
                            <NavLink className={({ isActive }) => (isActive ? 'active nav-link' : 'nav-link')} to='/contacts' onClick={hide}>
                                Contacts
                            </NavLink>
                            <NavLink className={({ isActive }) => (isActive ? 'active nav-link' : 'nav-link')} to='/cart' onClick={hide}>
                                {cartLabel}
                            </NavLink>
                        </div>
                    </div>
                </div>
            </nav>
            <div className='container pt-4'>
                <Routes>
                    <Route path='/' element={<Home />} />
                    <Route path='/catalog' element={<Catalog />} />
                    <Route path='/catalog/:id' element={<Product />} />
                    <Route path='/delivery' element={<Delivery />} />
                    <Route path='/contacts' element={<Contacts />} />
                    <Route path='/cart' element={<Cart />} />
                </Routes>
            </div>
        </div>
    );
};
