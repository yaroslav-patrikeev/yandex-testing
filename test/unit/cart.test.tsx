import '@testing-library/jest-dom';

import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { InternalAxiosRequestConfig } from 'axios';
import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import { CartApi, ExampleApi } from '../../src/client/api';
import { Cart } from '../../src/client/pages/Cart';
import { initStore } from '../../src/client/store';

const basename = '/hw/store';
const api = new ExampleApi(basename);
const cart = new CartApi();

describe('На странице корзина', () => {
	it('при инициализации происходит обращение к localStorage', async () => {
		const getState = jest.spyOn(cart, 'getState');
		const app = (
			<MemoryRouter initialEntries={[basename]} basename={basename}>
				<Provider store={initStore(api, cart)}>
					<Cart />
				</Provider>
			</MemoryRouter>
		);

		render(app);

		await waitFor(() => {
			expect(getState).toHaveBeenCalledTimes(1);
		});
	});
	it('должна отображаться таблица с добавленными в нее товарами', async () => {
		jest.spyOn(cart, 'getState').mockImplementation(() => {
			return { '0': { name: 'Unbranded kogtetochka', count: 1, price: 228 } };
		});
		const app = (
			<MemoryRouter initialEntries={[basename]} basename={basename}>
				<Provider store={initStore(api, cart)}>
					<Cart />
				</Provider>
			</MemoryRouter>
		);

		const { container } = render(app);

		await waitFor(() => {
			const table = container.querySelector('.Cart-Table');
			expect(table).toBeInTheDocument();
		});
	});
	it('в корзине должна быть кнопка "очистить корзину", по нажатию на которую все товары должны удаляться', async () => {
		jest.spyOn(cart, 'getState').mockImplementation(() => {
			return { '0': { name: 'Unbranded kogtetochka', count: 1, price: 228 } };
		});
		const app = (
			<MemoryRouter initialEntries={[basename]} basename={basename}>
				<Provider store={initStore(api, cart)}>
					<Cart />
				</Provider>
			</MemoryRouter>
		);

		const { container } = render(app);

		await waitFor(() => {
			const button = container.querySelector('.Cart button');
			expect(button).toBeInTheDocument();
		});
		const table = container.querySelector('.Cart-Table');
		expect(table).toBeInTheDocument();
		const button = container.querySelector('.Cart button');
		fireEvent.click(button);
		expect(table).not.toBeInTheDocument();
	});
	it('для каждого добавленного товара должны отображаться название, цена, количество , стоимость, а также должна отображаться общая сумма заказа в таблице', async () => {
		jest.spyOn(cart, 'getState').mockImplementation(() => {
			return { '0': { name: 'Unbranded kogtetochka', count: 1, price: 228 } };
		});
		const app = (
			<MemoryRouter initialEntries={[basename]} basename={basename}>
				<Provider store={initStore(api, cart)}>
					<Cart />
				</Provider>
			</MemoryRouter>
		);

		const { container } = render(app);

		await waitFor(() => {
			const table = container.querySelector('.Cart-Table');
			expect(table).toBeInTheDocument();
		});
		const [product, price, count, total] = container
			.querySelector('.Cart-Table tbody tr')
			.querySelectorAll('td') as NodeListOf<HTMLTableCellElement>;
		expect(product.textContent).toBe('Unbranded kogtetochka');
		expect(price.textContent).toBe('$228');
		expect(count.textContent).toBe('1');
		expect(total.textContent).toBe('$228');
	});
	it('если корзина пустая, должна отображаться ссылка на каталог товаров', async () => {
		const app = (
			<MemoryRouter initialEntries={[basename]} basename={basename}>
				<Provider store={initStore(api, cart)}>
					<Cart />
				</Provider>
			</MemoryRouter>
		);

		const { container } = render(app);

		await waitFor(() => {
			const link = container.querySelector('.Cart a');
			expect(link.textContent).toBe('catalog');
			expect(link).toBeInTheDocument();
		});
	});

	it('верстка в подтверждении отправки не изменилась', async () => {
		jest.spyOn(cart, 'getState').mockImplementation(() => {
			return { '0': { name: 'Unbranded kogtetochka', count: 1, price: 228 } };
		});
		jest.spyOn(api, 'checkout').mockImplementation(() => {
			return new Promise(res => {
				res({
					data: { id: 1 },
					status: 200,
					statusText: 'OK',
					headers: {},
					config: {
						headers: {},
					} as InternalAxiosRequestConfig,
				});
			});
		});
		const app = (
			<MemoryRouter initialEntries={[basename]} basename={basename}>
				<Provider store={initStore(api, cart)}>
					<Cart />
				</Provider>
			</MemoryRouter>
		);
		const { asFragment, container } = render(app);

		await waitFor(() => {
			const checkout = screen.getByText('Checkout');
			expect(checkout).toBeInTheDocument();
		});
		const inputs = container.querySelectorAll('.Form input');
		expect(inputs.length).toBe(2);
		const [name, phone] = inputs;
		const address = container.querySelector('.Form textarea');
		expect(address).toBeInTheDocument();

		fireEvent.change(name, { target: { value: 'Ivan' } });
		fireEvent.change(phone, { target: { value: '89999999999' } });
		fireEvent.change(address, { target: { value: 'my address' } });
		const button = container.querySelector('.Form button');
		expect(button).toBeInTheDocument();
		fireEvent.click(button);
		await waitFor(() => {
			const wellDone = container.querySelector('.alert-heading');
			expect(wellDone).toBeInTheDocument();
			expect(wellDone.textContent).toBe('Well done!');
			expect(asFragment()).toMatchSnapshot();
		});
	});
});

afterEach(() => {
	jest.restoreAllMocks();
});
