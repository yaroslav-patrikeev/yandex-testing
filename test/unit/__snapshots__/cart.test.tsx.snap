// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`На странице корзина верстка в подтверждении отправки не изменилась 1`] = `
<DocumentFragment>
  <div
    class="Cart"
  >
    <div
      class="row mb-4"
    >
      <div
        class="col"
      >
        <h1>
          Shopping cart
        </h1>
        <div
          class="row my-2"
        >
          <div
            class="col-12 col-sm-8 col-md-6"
          >
            <div
              class="Cart-SuccessMessage alert alert-success"
            >
              <h4
                class="alert-heading"
              >
                Well done!
              </h4>
              <p>
                Order #
                <strong
                  class="Cart-Number"
                >
                  1
                </strong>
                 has been successfully completed.
              </p>
              <hr />
              <p
                class="mb-0"
              >
                Please wait for confirmation of delivery.
              </p>
            </div>
          </div>
        </div>
        Cart is empty. Please select products in the 
        <a
          href="/hw/store/catalog"
        >
          catalog
        </a>
        .
      </div>
    </div>
  </div>
</DocumentFragment>
`;
